plugin/core/protocol.py

```diff
@@ -169,6 +171,42 @@ class SignatureHelpTriggerKind:
     'relatedInformation': List[DiagnosticRelatedInformation]
 }, total=False)
 
+TextEdit = TypedDict('TextEdit', {
+    'range': RangeLsp,
+    'newText': str
+}, total=True)
+
+InsertReplaceEdit = TypedDict('InsertReplaceEdit', {
+    'replace': RangeLsp,
+    'insert': RangeLsp,
+    'newText': str
+}, total=True)
+
+CompletionItem = TypedDict('CompletionItem', {
+    'label': str,
+    'kind': Optional[int],
+    'tags': Optional[List[int]],
+    'detail': Optional[str],
+    'documentation': Optional[Union[str, MarkupContent]],
+    'deprecated': Optional[bool],
+    'preselect': Optional[bool],
+    'sortText': Optional[str],
+    'filterText': Optional[str],
+    'insertText': Optional[str],
+    'insertTextFormat': Optional[int],
+    'insertTextMode': Optional[int],
+    'textEdit': Optional[Union[TextEdit, InsertReplaceEdit]],
+    'additionalTextEdits': Optional[List[TextEdit]],
+    'commitCharacters': Optional[List[str]],
+    'command': Optional[Command],
```

Comment:
The types here are not `Optional` I think.

Reply:
```

	/**
	 * A human-readable string that represents a doc-comment.
	 */
	documentation?: string | MarkupContent;
```

For the `documentation` key, the value type may be `string | MarkupContent`, 
but the `documentation?` indicates that that key might be missing in the object. (that is not the same as Optional in mypy)

I was not sure how to type hint that in mypy. 
There is `total=false`, but I see no clear way to type hint what keys might be missing. 
So for all those fields I added the `Optional` type wrapper to indicate that. (But that is not the correct interpenetration of the TypeScript `key?` type). 

Should I remove all those Optional wrapper types?

Reply:
Why do you think so?

In a case like this:

```py
c = {'label': 'foo'}  # type: CompletionItem
kind = c.get('kind')
```

what would you expect the type of `kind` be? I'd say `int` or `None` because `kind` is optional in the spec.

Reply:
There's a difference between optional keys and optional values... although the difference is hard to detect :)

Reply:
I would say that the difference is impossible to represent in Python types. To make those keys optional they would have to be missing completely from the Completion type which would then be impossible to know their type if they are present.

Reply:
> I would say that the difference is impossible to represent in Python types.

Not really true. You can detect the difference:

```python
# 1: optional keys
if "key" in obj:
    val = obj["key"]

# 2: optional values
val = obj["key"]
if val is not None:
    # ... handle val ...
```

Although that first (1) snippet does two dictionary lookups. It's faster to do one lookup:

```python
# 3: single lookup, but you cannot distinguish optional keys from optional values this way
val = obj.get("key")
if val is not None:
    # ... handle val ...
```

In any case, there are already usages of `total=False` exactly to tell mypy that some keys are missing. Although, with `total=False`, you cannot tell mypy which keys are mandatory. Either they are all mandatory or they are all optional. I thought I read somewhere that you can use the below trick to make a subset of the keys mandatory, and another subset of the keys optional:

```python
CompletionItem_Required(typing.TypedDict):
    label: str  # label is required

CompletionItem(CompletionItem_Required, total=False):
    kind: int,
    tags: List[int],
    detail: str   # kind, tags, detail, etc, are optional
```

Unfortunately, we cannot use this syntax as we're still on py33.

Moral of the story I think you can do one of two things:

1) Keep `total=False`, but then remove all the `Optional[...]`
2) Set `total=True`, and keep all the `Optional[...]`

Reply:
I see what you mean now.

> Moral of the story I think you can do one of two things:

Both of those options are unfortunate because then the person writing the code will not know which properties need to be checked for existence and which are expected to be always there.

In other words, the types specified as they are now, help the person writing the code write it in a way that doesn't crash without adding unnecessary checks for properties that are always there.

Reply:
We can introduce a new type 
```
OptionalKey = Optional
```

And later use it like:
```
'documentation': OptionalKey[str]
```

That way we have at least one way to distinguish an optional key, from an optional value. So a person writing the code would know what to do.

What do you think of this approach?